### ISSUE TITLE
ChiSquaredTest#goodness_of_fit hangs with BigDecimal input

### ISSUE THREAD

**Issue by chrisroos** [estebanz01/ruby-statistics#34]
This is the simplest example that seems to trigger the hanging behaviour. If the `BigDecimal` is cast to a float then it works as expected:

```
require 'statistics'

expected = [BigDecimal(0.2, 1)]
observed = [3.3]

chi_squared = Statistics::StatisticalTest::ChiSquaredTest
stats = chi_squared.goodness_of_fit(0.05, expected, observed)
# Observe that the process hangs at this point.
```

**Comment by estebanz01** [comment]
hey! Thanks for reporting this. I'll take a look and come back later :)

**Comment by estebanz01** [comment]
I cannot reproduce the issue with your example:
```ruby
[1] pry(main)> require 'bigdecimal'
=> true
[2] pry(main)> expected = [BigDecimal(0.2, 1)]
=> [0.2e0]
[3] pry(main)> observed = [3.3]
=> [3.3]
[4] pry(main)> 
[5] pry(main)> chi_squared = Statistics::StatisticalTest::ChiSquaredTest
=> Statistics::StatisticalTest::ChiSquaredTest
[6] pry(main)> stats = chi_squared.goodness_of_fit(0.05, expected, observed)
=> {:probability=>NaN, :p_value=>NaN, :alpha=>0.05, :null=>false, :alternative=>false, :confidence_level=>0.95}
```
But with this one, it "hangs":
```ruby
[7] pry(main)> observed_counts = [BigDecimal(212, 1), BigDecimal(147, 1), BigDecimal(103, 1), BigDecimal(50, 1), BigDecimal(46, 1), BigDecimal(42, 1)]
=> [0.212e3, 0.147e3, 0.103e3, 0.5e2, 0.46e2, 0.42e2]
[8] pry(main)> expected = BigDecimal(100,1)
=> 0.1e3
[9] pry(main)> chi_squared.goodness_of_fit(0.05, expected, observed_counts)
```

It hangs because it is more precise and it has to perform more compute expensive calculations like floating point calculation. Also, the values are platform dependant. I'll keep taking a look to see if there's a workaround to this.

Thanks for reporting it!

**Comment by estebanz01** [comment]
after half an hour, my example finished with this result:
```ruby
=> {:probability=>0.1000000000006968283386905586923867608561669020248017855044996415449374980061197643796005795577591392459319269458077753137619300911152053725168461181091036985331477e1,
 :p_value=>-0.6968283386905586923867608561669020248017855044996415449374980061197643796005795577591392459319269458077753137619300911152053725168461181091036985331477e-11,
 :alpha=>0.05,
 :null=>false,
 :alternative=>true,
 :confidence_level=>0.95}
```
Doing some logging, I found out that the issue is while performing one integral using the simpson's rule. Specifically, the [lower incomplete gamma](https://github.com/estebanz01/ruby-statistics/wiki/Helper-functions#lower-incomplete-gamma-function) function.

**Comment by estebanz01** [comment]
I'm wondering if this will gain some improvement with https://github.com/ruby/ruby/blob/master/doc/ractor.md

**Comment by estebanz01** [comment]
I have a hack for this, but not sure if it's worth it. That is, force all elements passed to the functions be `Float` rather than `BigDecimal` via `to_f`. I'm performing some experimentations around `BigDecimal` to see if I can fix it. The culprit is calculating the rectangles with `BigDecimal` https://github.com/estebanz01/ruby-statistics/blob/master/lib/math.rb#L48

**Comment by estebanz01** [comment]
@chrisroos hey! not sure if it's too late, but I introduced a change that allows the gem to work with `BigDecimal` by coercing to `Rational` and returning `BigDecimal` if the calculation allows it.

**Comment by estebanz01** [comment]
fixed on version 2.1.3.
